File: Lab-1/Example2-Part-BandC/MultithreadingExampleWorkers.java
import java.util.ArrayList;
import java.util.List;

public class MultithreadingExampleWorkers{


    static final int TOTAL_TASKS = 50;
    static final long WAIT_MS = 200;

    public static void main(String[] args) throws InterruptedException{

        int[] workerCounts = {1,2,3,4,5};

        for (int p : workerCounts) {
            
            long totalTime = 0;
            int repetitions = 2;

            for (int r = 0; r < repetitions; r++) {
                long start = System.nanoTime();
                runWithWorkers(p);
                long end = System.nanoTime();


                totalTime += (end - start);
            }

            double avgSeconds = (totalTime / (double) repetitions) / 1_000_000_000.0;

            System.out.printf("p=%d -> avg time: %.3f s%n", p, avgSeconds);
        }
    }


        static void runWithWorkers(int p) throws InterruptedException {

            List<Thread> workers = new ArrayList<>();
            int tasksPerWorker = TOTAL_TASKS / p;
            int remainder = TOTAL_TASKS % p;

            int assigned = 0;

            for (int i = 0; i < p; i++) {

                int tasksForThisWorker = tasksPerWorker + (i < remainder ? 1 : 0);
                int finalTasksForThisWorker = tasksForThisWorker;
                assigned += tasksForThisWorker;
 
                Thread t = new Thread(() -> {
                    for (int j = 0; j < finalTasksForThisWorker; j++) {
                        performTask();
                    }
                });

                workers.add(t);
            }

            for (Thread t : workers) t.start();
            for (Thread t : workers) t.join();


        }

        static void performTask() {

            try {

                Thread.sleep(WAIT_MS);

            } catch (InterruptedException e) {

                Thread.currentThread().interrupt();
            }
        }
}

    

File: Lab-1/Example2-Part-BandC/MultithreadingExampleWorkers.cpp
#include <iostream>
#include <thread>
#include <chrono>
#include <vector>

// to compile g++ -std=c++20 -O2 -pthread MultithreadingExampleWorkers.cpp -o MultithreadingExampleWorkers

const int TOTAL_TASKS = 50;
const long WAIT_MS = 200;

void performTask() {
    std::this_thread::sleep_for(std::chrono::milliseconds(WAIT_MS));
}

void runWithWorkers(int p) {
    std::vector<std::thread> workers;

    int tasksPerWorker = TOTAL_TASKS / p;
    int remainder = TOTAL_TASKS % p;

    for (int i = 0; i < p; ++i) {
        int tasksForThisWorker = tasksPerWorker + (i < remainder ? 1 : 0);

        workers.emplace_back([tasksForThisWorker]() {
            for (int j = 0; j < tasksForThisWorker; ++j) {
                performTask();
            }
        });
    }

    for (auto& t : workers) t.join();
}

int main() {
    std::vector<int> workerCounts = {1, 2, 3, 4, 5};
    int repetitions = 2;

    for (int p : workerCounts) {
        double totalTime = 0.0;

        for (int r = 0; r < repetitions; ++r) {
            auto start = std::chrono::steady_clock::now();
            runWithWorkers(p);
            auto end = std::chrono::steady_clock::now();

            std::chrono::duration<double> elapsed = end - start;
            totalTime += elapsed.count();
        }

        double avgSeconds = totalTime / repetitions;
        std::cout << "p=" << p << " -> avg time: " << avgSeconds << " s" << std::endl;
    }

    return 0;

}
